src/SpirvIR.h

```diff
@@ -0,0 +1,557 @@
+#ifndef HALIDE_SPIRV_IR_H
+#define HALIDE_SPIRV_IR_H
+
+/** \file
+ * Defines methods for constructing and encoding instructions into the Khronos
+ * format specification known as the Standard Portable Intermediate Representation
+ * for Vulkan (SPIR-V). These class interfaces adopt Halide's conventions for its
+ * own IR, but is implemented as a stand-alone optional component that can be
+ * enabled as required for certain runtimes (eg Vulkan)
+ */
+#ifdef WITH_SPIRV
+
+#include <map>
+#include <set>
+#include <stack>
+#include <unordered_map>
+#include <vector>
+
+#include "IntrusivePtr.h"
+#include "Type.h"
+
+#include <spirv/1.0/spirv.h>  // Use v1.0 spec as the minimal viable version (for maximum compatiblity)
```

Comment:
This will end up in the user-supplied `Halide.h` which means consumers will need to have the SPIR-V headers available. Is this _necessary_? If so, new code will need to be written in our package config file to handle creating the `Halide::SPIRV` target. That likely means writing a Find module that calls `find_path` the same way I do below.

If this `#include` can be confined to `SpirvIR.cpp`, then the situation is simpler.

Reply:
I'll go further and say that it **cannot** be necessary. We must not require these headers for Halide.h.

Reply:
Hrmmm ... we only include a single stand-alone header file which is the SPIR-V typedefs and enums, so there's no additional dependencies.

Reply:
Specifically, this file: [spirv.h](https://github.com/KhronosGroup/SPIRV-Headers/blob/master/include/spirv/1.0/spirv.h)

Reply:
This seems require an external header from a thirdparty to be present to compile Halide. The obvious solution to that is to add the file to this PR. It also has a different license, which should not be a problem, but... it kinda always is... I'd also look at putting this code in a subdirectory. I know we have not done so for other stuff but it is getting to be time to do that. It may be enough to just add the Khronos header to a thirdparty directory and include it from there. (We have done some things like this for the BitWriter support and the NVidia ptx libs.)

I do agree that maybe its time to have some stuff in the compiler that isn't in Halide.h...

Reply:
This would certainly simplify things. I can add the header file to the dependencies/spirv folder along with the Khronos license if that works.

Reply:
Why is this going to end up in Halide.h though? Can we just keep SpirvIR.h out of Halide.h?

Reply:
Absolutely ... the current use of SpirvIR.h is only going to be inside of CodeGen_Vulkan.cpp's implementation (forthcoming).  If there's a better way to organize this to make this clear, I'm more than happy to restructure things!

Reply:
ok, so we just need to take it out of the list of headers that makes up Halide.h, and it'll be fine. Looking at the change to the Makefile, it's currently in the list. I didn't check the cmake build.